validate.go

```diff
@@ -16,36 +17,133 @@ func validate(payload []byte) ([]byte, error) {
+		"request.object.spec.volumes")
+	if !volumes.Exists() {
+		// pod defines no volumes, accepting
 		return kubewarden.AcceptRequest()
 	}
-	name := data.String()
 
-	logger.DebugWithFields("validating ingress object", func(e onelog.Entry) {
-		namespace := gjson.GetBytes(payload, "request.object.metadata.namespace").String()
+	logger.DebugWithFields("validating pod object", func(e onelog.Entry) {
+		name := gjson.GetBytes(payload, "request.object.metadata.name").String()
+		namespace := gjson.GetBytes(payload,
+			"request.object.metadata.namespace").String()
 		e.String("name", name)
 		e.String("namespace", namespace)
 	})
 
-	if settings.DeniedNames.Contains(name) {
-		logger.InfoWithFields("rejecting ingress object", func(e onelog.Entry) {
+	// build list of volumeMounts from all the initContainers and containers:
+	volumeMounts := make([]gjson.Result, 0)
+	cases := []string{
+		"request.object.spec.initContainers",
+		"request.object.spec.containers",
+	}
+	for _, c := range cases {
+		containers := gjson.GetBytes(payload, c)
+		for _, container := range containers.Array() {
+			containerVolumeMounts := gjson.Get(container.String(),
+				"volumeMounts")
+			for _, mount := range containerVolumeMounts.Array() {
+				volumeMounts = append(volumeMounts, mount)
+			}
+		}
+	}
+
+out: // label for skipping all fors
+	for _, volume := range volumes.Array() {
+		for _, mount := range volumeMounts {
+			if gjson.Get(volume.String(), "hostPath").Exists() {
+				// volume is of type hostPath
+
+				// if volumeMount object matches the 'volume.name':
+				if gjson.Get(volume.String(), "name").String() == /* mount name */
+					gjson.Get(mount.String(), "name").String() { /* volume name */
+					// volume is a hostpath and it in use, validate against
+					// settings
+
+					readOnly := false // volumeMount.readOnly missing means 'false'
+					if gjson.Get(mount.String(), "readOnly").Exists() {
+						readOnly = gjson.Get(mount.String(), "readOnly").Bool()
+					}
+
+					path := gjson.Get(volume.String(), "hostPath.path").String()
+
+					match := false
+					// readOnly attribute of most specific AllowedHostPath takes precendence:
+					previousAllowedHostPath := ""
+					for _, allowedHostPath := range settings.AllowedHostPaths {
+						if hasPathPrefix(path, allowedHostPath.PathPrefix) {
+							// current setting allowedHostPath matches path of volumeMount
+							if hasPathPrefix(allowedHostPath.PathPrefix, previousAllowedHostPath) {
+								// allowedHostPath is more specific (and has precendence over
+								//	past allowedHostPath), or the same path
+								match = true
+								err = validatePath(path, readOnly, allowedHostPath)
+								previousAllowedHostPath = allowedHostPath.PathPrefix
+							}
+						}
+					}
+					if !match {
+						// path didn't match against any PathPrefix in settings
+						err = fmt.Errorf("hostPath '%s' is not in the AllowedHostPaths list",
+							path)
```

Comment:
Can we build up errors, like we do during settings validation, so that an user gets back the list of **all** the host paths that violate the policy?

Reply:
Done in https://github.com/kubewarden/psp-hostpaths-policy/pull/1/commits/eac35413782c7bddfc8078c8cc0d53445ae83a29, https://github.com/kubewarden/psp-hostpaths-policy/pull/1/commits/3b04f997caf79034366b56d2ed8d391173394431 and tests in https://github.com/kubewarden/psp-hostpaths-policy/pull/1/commits/1e2c0b17b747f56d61439bf6004e9d932d59acb7, https://github.com/kubewarden/psp-hostpaths-policy/pull/1/commits/d6eb2ba334237c44c5d9933e668c04815ab345ab.

Reply:
That's totally fine, no worries